### PR TITLE
Refactor ScriptType to be a Top-Level Class

### PR DESCRIPTION
**Pull request by jdconrad**
As the title says this is the only change.  I'm going to attempt to clean up some of the scripting API, and this is the first in a series of smaller changes moving forward.  The lines of code appear much larger than the change itself simply because so many import statements changed.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/script/ScriptType.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    FILE(2, "file", "file", true);
+
+    private final int val;
+    private final ParseField parseField;
+    private final String scriptType;
+    private final boolean defaultScriptEnabled;
+
+    public static ScriptType readFrom(StreamInput in) throws IOException {
+        int scriptTypeVal = in.readVInt();
+        for (ScriptType type : values()) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/script/ScriptType.java]: does it make sense to do this with a switch / case and throw in default?

File: core/src/main/java/org/elasticsearch/script/ScriptType.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            if (type.val == scriptTypeVal) {
+                return type;
+            }
+        }
+        throw new IllegalArgumentException("Unexpected value read for ScriptType got [" + scriptTypeVal + "] expected one of ["
+                + INLINE.val + "," + FILE.val + "," + STORED.val + "]");
+    }
+
+    public static void writeTo(ScriptType scriptType, StreamOutput out) throws IOException{
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/script/ScriptType.java]: do we have to allow null values? there is also a space missing after `IOException`. Can you, if we allow null at least document this?

File: core/src/main/java/org/elasticsearch/script/ScriptType.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        this.parseField = new ParseField(name);
+        this.scriptType = scriptType;
+        this.defaultScriptEnabled = defaultScriptEnabled;
+    }
+
+    public ParseField getParseField() {
+        return parseField;
+    }
+
+    public boolean getDefaultScriptEnabled() {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/script/ScriptType.java]: `isDefaultScriptEnabled`?

**Review by s1monw — COMMENTED** [review]
LGTM in general, I left some minors. I think we can also target 5.1 for this

**Comment by jdconrad** [comment]
@s1monw I literally just moved the class from it's current existing state to top-level.  I did not touch the class itself.  I have some planned refactoring specifically for this class in my next set of changes.  Would it be all right if I covered the comments you mentioned here (all things I agree with completely :) ) in the next PR?

**Comment by s1monw** [comment]
sure

**Comment by jdconrad** [comment]
@s1monw Awesome, thank you!

**Comment by dadoonet** [comment]
Note that this is a breaking change for the Java API as people can't use do anymore:

```java
        SearchTemplateResponse str = new SearchTemplateRequestBuilder(client)
                .setScript("template_gender")
                .setScriptType(ScriptService.ScriptType.STORED)
                .setRequest(new SearchRequest())
                .setScriptParams(template_params)
                .get();
``` 

Not a big deal though. I'm just going to update our Java API documentation: https://www.elastic.co/guide/en/elasticsearch/client/java-api/5.1/java-search-template.html

**Pull request merged**